File: src/main/scala/actor/DatabaseReaderActor.scala
package actor

import scala.slick.driver.SQLiteDriver.simple._
import util.Configuration
import akka.actor.Actor

case class DatabaseReaderActor(databaseName: String) extends Actor {

  private lazy val additionalFiles = Configuration.additionalFiles

  val driver = "org.sqlite.JDBC"

  private lazy val database = Database.forURL(databaseUrl(databaseName), driver = driver)

  private var session: Session = null

  override def preStart() {
    super.preStart()
    session = database.createSession()
  }

  override def postStop() {
    super.postStop()
    if (session != null) {
      session.close()
    }
  }

  override def preRestart(reason: Throwable, message: Option[Any]) {
    super.preRestart(reason, message)
    if (session != null) {
      session.close()
    }

  }

  override def postRestart(reason: Throwable) {
    super.postRestart(reason)
    session = database.createSession()
  }

  def receive = {
    case Extract(f) => {

      database.withTransaction {
        session:Session => {
          val result = f(session)

          sender ! Extracted(result)
        }
      }
    }
  }

  def databaseUrl(databaseName: String): String = {
    "jdbc:sqlite://%s/%s".format(additionalFiles, databaseName)
  }
}

File: src/main/java/actor/DatabaseReaderActor.java
package actor;

import akka.actor.UntypedActor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.SimpleDriverDataSource;
import util.Configuration;

import javax.sql.DataSource;
import java.io.File;
import java.net.URISyntaxException;
import java.sql.Driver;
import java.sql.DriverManager;
import java.sql.SQLException;

public class DatabaseReaderActor extends UntypedActor {

    private final String databaseName;

    private static final String DRIVER = "org.sqlite.JDBC";

    private JdbcTemplate template;


    public DatabaseReaderActor(String databaseName) {
        this.databaseName = databaseName;
    }

    @Override
    public void onReceive(Object message) throws Exception {
        //TODO: IMPLEMENTS JDBC QUERIES
    }

    @Override
    public void preStart() {
        String path = Configuration.additionalFiles;
        try {
            File directory = new File(getClass().getClassLoader().getResource(path).toURI());
            Class.forName(DRIVER);
            String url = String.format("jdbc:sqlite://%s/%s", directory.getAbsolutePath(), databaseName);
            Driver driver = DriverManager.getDriver(url);
            DataSource dataSource = new SimpleDriverDataSource(driver, url);
            template = new JdbcTemplate(dataSource);

            if (!directory.isDirectory()) {
                throw new IllegalStateException(String.format("%s must exists", path));
            }

        } catch (URISyntaxException e) {
            throw new IllegalStateException(String.format("%s must exists", path));
        } catch (ClassNotFoundException e) {
            throw new IllegalStateException(e);
        } catch (SQLException e) {
            throw new IllegalStateException(e);
        }

    }


}
